test/integration/authorization_test.go

```diff
@@ -609,55 +612,205 @@ type subjectAccessReviewTest struct {
+		failMessage := ""
+		err := wait.Poll(testutil.PolicyCachePollInterval, testutil.PolicyCachePollTimeout, func() (bool, error) {
+			var err error
+			var actualResponse kubeauthorizationapi.SubjectAccessReviewStatus
 			if test.localReview != nil {
-				failMessage = fmt.Sprintf("%s: from local review\n\t%#v\nexpected\n\t%#v\ngot\n\t%#v", test.description, test.localReview, &test.response, actualResponse)
+				if len(test.localReview.User) == 0 && (test.localReview.Groups == nil || len(test.localReview.Groups.UnsortedList()) == 0) {
+					var tmp *kubeauthorizationapi.SelfSubjectAccessReview
+					if tmp, err = test.kubeAuthInterface.SelfSubjectAccessReviews().Create(toKubeSelfSAR(testNS, test.localReview)); err == nil {
+						actualResponse = tmp.Status
+					}
+				} else {
+					var tmp *kubeauthorizationapi.LocalSubjectAccessReview
+					if tmp, err = test.kubeAuthInterface.LocalSubjectAccessReviews(testNS).Create(toKubeLocalSAR(testNS, test.localReview)); err == nil {
+						actualResponse = tmp.Status
+					}
+				}
 			} else {
-				failMessage = fmt.Sprintf("%s: from review\n\t%#v\nexpected\n\t%#v\ngot\n\t%#v", test.description, test.clusterReview, &test.response, actualResponse)
+				var tmp *kubeauthorizationapi.SubjectAccessReview
+				if tmp, err = test.kubeAuthInterface.SubjectAccessReviews().Create(toKubeClusterSAR(test.clusterReview)); err == nil {
+					actualResponse = tmp.Status
+				}
 			}
-			return false, nil
+			testErr := test.kubeErr
+			if len(testErr) == 0 {
+				testErr = test.err
+			}
+			if len(testErr) > 0 {
+				if err == nil {
+					failMessage = fmt.Sprintf("%s: Expected error: %v\ngot\n\t%#v", test.description, testErr, actualResponse)
+					return false, nil
+				} else if !strings.HasPrefix(err.Error(), testErr) {
+					failMessage = fmt.Sprintf("%s: expected\n\t%v\ngot\n\t%v", test.description, testErr, err)
+					return false, nil
+				}
+			} else {
+				if err != nil {
+					failMessage = fmt.Sprintf("%s: unexpected error: %v", test.description, err)
+					return false, nil
+				}
+			}
+
+			if (actualResponse.Allowed != test.response.Allowed) || (!strings.HasPrefix(actualResponse.Reason, test.response.Reason)) {
+				if test.localReview != nil {
+					failMessage = fmt.Sprintf("%s: from local review\n\t%#v\nexpected\n\t%#v\ngot\n\t%#v", test.description, test.localReview, &test.response, actualResponse)
+				} else {
+					failMessage = fmt.Sprintf("%s: from review\n\t%#v\nexpected\n\t%#v\ngot\n\t%#v", test.description, test.clusterReview, &test.response, actualResponse)
+				}
+				return false, nil
+			}
+
+			failMessage = ""
+			return true, nil
+		})
+
+		if err != nil {
+			t.Error(err)
 		}
+		if len(failMessage) != 0 {
+			t.Error(failMessage)
+		}
+	}
+}
 
-		failMessage = ""
-		return true, nil
-	})
+func toKubeSelfSAR(testNS string, sar *authorizationapi.LocalSubjectAccessReview) *kubeauthorizationapi.SelfSubjectAccessReview {
```

Comment:
Are these fully convertible to and from?  I'll give this to you for now, but if they are convertible, you should write a real conversion and fuzz it.